test/markup.c

```diff
@@ -45,12 +45,97 @@ TEST test_markup_transform(void)
         ASSERT_STR_EQ("<i>foo</i> bar baz", (ptr=markup_transform(g_strdup("<i>foo</i><br>bar\nbaz"), MARKUP_FULL)));
         g_free(ptr);
 
+        // Test replacement of img and a tags, not renderable by pango
+        ASSERT_STR_EQ("foo bar bar baz", (ptr=markup_transform(g_strdup("<img alt=\"foo bar\"><br>bar\nbaz"), MARKUP_FULL)));
+        g_free(ptr);
+        ASSERT_STR_EQ("test ",           (ptr=markup_transform(g_strdup("test <img alt=\"foo bar\""), MARKUP_FULL)));
+        g_free(ptr);
+        ASSERT_STR_EQ("test  image",     (ptr=markup_transform(g_strdup("test <img src=\"nothing.jpg\"> image"), MARKUP_FULL)));
+        g_free(ptr);
+        ASSERT_STR_EQ("bar baz",         (ptr=markup_transform(g_strdup("<a href=\"asdf\">bar</a> baz"), MARKUP_FULL)));
+        g_free(ptr);
+
+        PASS();
+}
+
+TEST helper_markup_strip_a (const char *in, const char *exp, const char *urls)
+{
+        char *out = g_strdup(in);
+        char *out_urls = NULL;
+        char *msg = g_strconcat("url: ", in, NULL);
+
+        markup_strip_a(&out, urls ? &out_urls : NULL);
+
+        ASSERT_STR_EQm(msg, exp, out);
+        ASSERT_STR_EQm(msg, urls ? urls : "", out_urls ? out_urls : "");
+
+        g_free(out_urls);
+        g_free(out);
+        g_free(msg);
+
+        PASS();
+}
+
+TEST test_markup_strip_a(void)
+{
+        RUN_TESTp(helper_markup_strip_a, "<a href=\"https://url.com\">valid</a> link",   "valid link", "[valid] https://url.com");
+        RUN_TESTp(helper_markup_strip_a, "<a href=\"\">valid</a> link",                  "valid link", "[valid] ");
+        RUN_TESTp(helper_markup_strip_a, "<a>valid</a> link",                            "valid link", NULL);
+        RUN_TESTp(helper_markup_strip_a, "<a href=\"https://url.com\">valid link",       "valid link", "[valid link] https://url.com");
+
+        RUN_TESTp(helper_markup_strip_a, "<a href=\"https://url.com\" invalid</a> link", " link",      NULL);
+        RUN_TESTp(helper_markup_strip_a, "<a invalid</a> link",                          " link",      NULL);
+
+        PASS();
+}
+
+TEST helper_markup_strip_img (const char *in, const char *exp, const char *urls)
+{
+        char *out = g_strdup(in);
+        char *out_urls = NULL;
+
+        char *msg = g_strconcat("url: ", in, NULL);
+
+        markup_strip_img(&out, urls ? &out_urls : NULL);
+
+        ASSERT_STR_EQm(msg, exp, out);
+        ASSERT_STR_EQm(msg, out_urls ? out_urls : "", urls ? urls : "");
+
+        g_free(out_urls);
+        g_free(out);
+        g_free(msg);
+
+        PASS();
+}
+
+TEST test_markup_strip_img(void)
+{
+        RUN_TESTp(helper_markup_strip_img, "v <img> img",                                         "v  img",       NULL);
```

Comment:
Can tests really be nested like that? I don't see anything indicating so in `greatest` docs.

Reply:
Yes and no. The code is indicating it, but it's nowhere documented.

It works and it's no problem. In functionality it's far better than calling an extra function.

I wish, there would be an additional layer between test and assertion.

Reply:
Ok a small background info on this: When an assertion fails, it immediately returns the current function (which should be a test).

As we test multiple strings (done via multiple assertions), a failed assertion breaks the test function. By calling another subtest, the supertest does not fail and will test further for all strings.